Src/FluentAssertions/Formatting/AttributeBasedFormatter.cs

```diff
@@ -46,9 +46,34 @@ public string Format(object value, FormattingContext context, FormatChild format
         private MethodInfo GetFormatter(object value)
         {
             Type valueType = value.GetType();
-            MethodInfo formatter = Formatters.FirstOrDefault(m => m.GetParameters().Single().ParameterType == valueType);
+            var possibleFormatters = Formatters
+                .Select(m => new
+                {
+                    Type = m.GetParameters().Single().ParameterType,
+                    Method = m
+                })
+                .Where(f => valueType.IsSameOrInherits(f.Type))
+                .ToDictionary(f => f.Type, f => f.Method);
```

Comment:
🤔 Is there a need to order the formatters in some way to get some deterministic order?

Reply:
That would require multiple enumerations and I guess wouldn't improve readability.
One enumeration using do/while looks more efficient and readable to me.

Reply:
Imagine multiple classes implementing the same interface `IFoo`. What would happen if you two formatters, one that accepts a `IFoo` and one, more specialized, that takes an implementation of `IFoo`.

Reply:
@Serg046 if efficiency is a concern here, could you then cache the `Dictionary` just as `formatters` is?

Reply:
Did it but forced to change one existing test a bit because of cache.